### DISCUSSION TITLE
Idea for supporting user-defined macro in LaTeX

### DISCUSSION THREAD

**Discussion by fzyxh** [Abhi-1U/texor#74] [General]
To support user-defined macro packages, maybe the `standalone` package can help. More detail in [this page](https://tex.stackexchange.com/questions/395405/is-it-possible-to-usepackage-inside-an-input-or-include-file).

**Comment by hturner** [comment]
Is the issue the way that the R Journal uses a wrapper for the main .tex file? I think this is a complication we do not need for a general .Rnw to .Rmd converter, because we do not need to add all the journal metadata (volume and publication date etc). We can just have one .tex file with all the front matter from the original article.

**Reply by fzyxh** [reply]
Using wrapper is not a necessary step for converter so far. The biggest problem is that most original functions can only pass one parameter (directory). I use it just because I can reuse most of the original functions. If we deprecate the wrapper, we should also modify original functions.

**Comment by Abhi-1U** [comment]
Pandoc does not use LaTeX engine to convert the documents, it is more like an markup convertor which will only parse a subset of LaTeX syntax. So when the style files are included in the article pandoc will go through the latex code in the style file and try to parse the logic on its own, but the catch is : 
1.  The style file needs to be in the same directory as the article (Hence the Metafix.sty file is always added to the working directory)
2. Pandoc's LaTeX reader can comprehend simple macros under `\newcommand{}` but as soon as we start using packages with lower level TeX/LaTeX  syntax, pandoc will ignore it.

So it actually does not matter if the we include the whole style file, rather port over essential macros to the existing(Metafix.sty) style file or create a new one base on the vignette style file for sweave files. 

 Also while researching about this I found out an alternative way to make pandoc do the substitutions in the math env as well, it is worth giving a try (https://pandoc.org/MANUAL.html#extension-latex_macros). Extensions are added to the input format [here](https://github.com/fzyxh/texor/blob/2737bdb7a2eb31d469ec5e2f4ce335ad4d919a89/R/article-tools.R#L152) similar to the one I have done for the output format.

**Reply by fzyxh** [reply]
I tried this method, but the output doesn't what I expected.

It just preserve macros in R Markdown file, and parse them when knit into HTML. The output R Markdown file is like:

````Markdown
---
title: Comparing Least Squares Calculations
abstract: |
  Many statistics methods require one or more least squares problems to
  be solved. There are several ways to perform this calculation, using
  objects from the base R system and using objects in the classes
  defined in the `Matrix` package.

  We compare the speed of some of these methods on a very small example
  and on a example for which the model matrix is large and sparse.
author: |
  Douglas Bates\
  R Development Core Team\
  [Douglas.Bates@R-project.org](Douglas.Bates@R-project.org){.uri}
date: '2024-06-01'
packages:
  cran: []
  bioc: []
vignette: |-
  %\VignetteEngine{knitr::knitr}
  %\VignetteEncoding{UTF-8}
  %\VignetteIndexEntry{Comparisons of Least Squares calculation speeds}
output: rmarkdown::html_vignette

---

\RequirePackage{hyperref}
\RequirePackage{url}
\RequirePackage{amsmath}
\RequirePackage{bm}%-> \bm (= bold math)
...
\newcommand{\bX}{\ensuremath{\bm{X}}}
...
\newcommand{\bzer}{\ensuremath{\bm{0}}}

::: article
``` {r include=FALSE}
library(knitr)
opts_chunk$set(
engine='R',dev='pdf',fig.width=5,fig.height=3,strip.white=TRUE,tidy=FALSE
)
```

``` {r preliminaries, echo=FALSE}
options(width=75)
library(stats) # for R_DEFAULT_PACKAGES=NULL
library(utils) # ditto
```

# Linear least squares calculations {#sec:LeastSquares}

Many statistical techniques require least squares solutions
$$\label{eq:LeastSquares}
  \widehat{\mathbf{\beta}}=
  \arg\min_{\mathbf{\beta}}\left\|\mathbf{y}-\bX\mathbf{\beta}\right\|^2   (\#eq:LeastSquares)$$
where $\bX$ is an $n\times p$ model matrix ($p\leq n$), $\mathbf{y}$ is
$n$-dimensional and $\mathbf{\beta}$ is $p$ dimensional. Most statistics
texts state that the solution to (\@ref(eq:LeastSquares)) is
$$\label{eq:XPX}
  \widehat{\mathbf{\beta}}=\left(\bX\trans\bX\right)^{-1}\bX\trans\mathbf{y}   (\#eq:XPX)$$
when $\bX$ has full column rank (i.e. the columns of $\bX$ are linearly
independent) and all too frequently it is calculated in exactly this
way.

...

:::

````

I put two HTML output screenshots from R Markdown before and after using latex_macro below.

`\bm` can works in latex, but pandoc and R Markdown don't know what it is. So using latex_macro cannot solve the problem completely, and it makes a mess of R Markdown file.

![image](https://github.com/Abhi-1U/texor/assets/20641413/966814a3-de90-4a4a-8084-04aa9cd9337b)

![image](https://github.com/Abhi-1U/texor/assets/20641413/bc0a2e8d-e54a-4886-bd58-33db6a5c061f)

We use another approach that requires two actions. On the one hand, we should add new macros into `Metafix.sty` like `\newcommand{\bm}[1]{\boldsymbol{\mathbf{#1}}}`. On the other hand, we should include user-defined macros into `RJwrapper.tex` like `\usepackage{myVignette}`. We can get the right result in the end as screenshot below (ignore references for equation plz):

![image](https://github.com/Abhi-1U/texor/assets/20641413/56a1739a-bdf6-43e1-b3cb-6f2ae353e873)

To compare with original pdf version, here's pdf vignette screenshot below: 

![image](https://github.com/Abhi-1U/texor/assets/20641413/6100c34c-5a7e-459b-8f81-ed25a80a0c2b)

**Reply by fzyxh** [reply]
Now you can try with latest version of [fzyxh/texor](https://github.com/fzyxh/texor/tree/6fa98877262c157c8860ac8215cbd843a083bb65).

`Comparisons.Rnw` as an example, please make sure `myVignette.sty` is in the same directory:

```
.        # current directory
|-- Comparisons_mod.Rnw
|-- myVignette.sty
```

Execute `rnw_to_rmd()` function with latest branch and you will get the output the same as the last screenshot above.

texor now automatically detects **included self-defined macros in the same directory** with `.Rnw` file, and adds them into `RJwrapper.tex`

**Reply by Abhi-1U** [reply]
1. Yes `\bm` is not supported in MathJax hence we have to substitute it with `\boldmath{#1}` or other equivalent using a macro.
2. The Style file needs included in the wrapper file, currently I have added it manually but we would need to craft a different wrapper file for sweave file or make changes in the functions to work without the wrapper. 
3. I have uploaded the same in my fork of texor playground [commit](https://github.com/Abhi-1U/texor-playground/commit/0b337cd2ada456b1bbc54352acc98e0414cf9a90). Its a crude example, I think we might not need to modify the  `myVignette` style file  much apart from the `\bm` which you might already have done.

**Reply by fzyxh** [reply]
Separating wrapper for different format output is a good idea. I think we can do this after the workflow of `rnw_to_rmd` works well.
